crowbar_framework/lib/upgrade.rb

```diff
@@ -0,0 +1,64 @@
+class Upgrade
+  attr_accessor :current_step, :substep, :current_node, :upgrade_progress
+  attr_reader :steps
+
+  # Return the current state of upgrade process.
+  # We're keeping the information in the file so is accessible by other applications
+  # and different crowbar versions.
+  def initialize
+    @upgrade_progress = {
+      current_step: upgrade_steps_6_7.first,
+      # substep is needed for more complex steps like upgrading the nodes
+      substep: "",
+      # current node is relevant only for the nodes_upgrade step
+      current_node: ""
+    }
+    if progress_file_path.exist?
+      @upgrade_progress = JSON.load(progress_file_path.read)
+    else
+      # in 'steps', we save the information about each step that was executed
+      @upgrade_progress["steps"] = upgrade_steps_6_7.map do |step|
+        [step, { status: "pending", errors: {} }]
+      end.to_h
+    end
+    @current_step = @upgrade_progress["current_step"]
+    @current_substep = @upgrade_progress["current_substep"]
+    @current_node = @upgrade_progress["current_node"]
+    @steps = @upgrade_progress["steps"]
+  end
+
+  def save
+    @upgrade_progress["current_step"] = @current_step
+    @upgrade_progress["current_substep"] = @current_substep
+    @upgrade_progress["current_node"] = @current_node
+    progress_file_path.open("w") do |f|
+      f.write(JSON.pretty_generate(@upgrade_progress))
+    end
+  end
+
+  def finished?
+    @upgrade_progress["current_step"] == upgrade_steps_6_7.last
+  end
+
+  protected
+
+  # global list of the steps of the upgrade process
+  def upgrade_steps_6_7
```

Comment:
This doesn't quite feel right to me - it's hardcoding an assumption that the top-level steps are linear.  OTOH it's possible that for this one-off upgrade case that's a safe assumption, so I suppose we could live with that and then adapt later if needed.